keps/sig-api-machinery/2558-field-version-openapi/README.md

```diff
@@ -0,0 +1,414 @@
+This will remove the `x-kubernetes-field-version` vendor extension for
+the field in OpenAPI.
+
+<<[UNRESOLVED]>>
+
+Open questions from @thockin's proposal:
+
+1) Should the tooling derive the prerelease status by looking up the gate name
+(one place to assert).
+
+2) Should we also tag structs or even new Kinds when they are gated?  This is
+less common and less well-defined.
+
+3) Should we tag *all* fields, even those that are not gated, so that tooling
+can know that the absence of the tag is an error, rather than indicating GA.
+
+<<[/UNRESOLVED]>>
+
+### Example
+
+Consider adding a new alpha API field `Width` to the stable API `Frobber`
+in `v1.20`.
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // This field is alpha-level and is only honored by servers that enable the Frobber2D feature.
+  // +optional
+  Width  *int32 `json:"width,omitempty"
+}
+```
+
+The field `Width` is controlled by the `Frobber2D` feature gate defined
+in [`kube_features.go`].
+
+```go
+// owner: @you
+// alpha: v1.20
+//
+// Add multiple dimensions to frobbers.
+Frobber2D utilfeature.Feature = "Frobber2D"
+
+var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureSpec{
+  ...
+  Frobber2D: {Default: false, PreRelease: utilfeature.Alpha},
+}
+```
+
+To link the feature gate with the API field, add the following comment line
+with the `+k8s:openapi-gen` prefix for the `Width` field:
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // +optional
+  // +k8s:openapi-gen:prerelease=alpha,minVersion=v1.20,featureGate=Frobber2D
+  Width  *int32 `json:"width,omitempty"
+}
+```
+
+**Note**: the comments on the feature gate are not meant to be machine parseable.
+They may diverge from the comments on the API field.
+The source of truth for the versioning information are
+the comments on the API field. See [Caveats](#caveats) section below.
+
+This will generate an OpenAPI schema with an additional vendor extension field:
+
+```json
+"width": {
+  VendorExtensible: spec.VendorExtensible{
+    Extensions: spec.Extensions{
+      "x-kubernetes-field-version": []interface{}{
+        map[string]interface{}{
+          "prerelease":  "alpha",
+          "minVersion":  "v1.20",
+          "featureGate": "Frobber2D",
+        },
+      },
+    },
+  },
+},
+```
+
+This shows that the field is alpha from v1.20 and 
+is controlled by the `Frobber2D` feature gate.
+
+### Constraints and Caveats
+
+#### Constraints
+
+The following constraints are applied to the comment tags
+and validated:
+
+- The generator would only consider comment lines with the
+`+k8s:openapi-gen:prerelease=` prefix i.e. the `prerelease`
+part is mandatory.
+- If a comment line starts with `+k8s:openapi-gen:prerelease=`,
```

Comment:
Not sure if this sentence means that they are mandatory?

Reply:
Clarified :+1: